redis/src/tls.rs

```diff
@@ -0,0 +1,160 @@
+use std::io::{BufRead, Error, ErrorKind as IOErrorKind};
+
+use rustls::{Certificate, OwnedTrustAnchor, PrivateKey, RootCertStore};
+use rustls_pemfile::{certs, read_one};
+
+use crate::{Client, ConnectionAddr, ErrorKind, IntoConnectionInfo, RedisError, RedisResult};
+
+/// The builder takes all parameters necessary to create a TLS connection.
+///
+/// In addition to an `url` with `rediss` location, it must be provided with
+/// binary streams corresponding to PEM file's contents for:
+/// - CA's certificate
+/// - one client certificate and
+/// - client's key
+///
+/// Example usage:
+///
+/// ```no_run
+/// use redis::{self, build_with_tls};
+/// use redis::AsyncCommands;
+///
+/// use std::fs::File;
+/// use std::io::BufReader;
+///
+/// async fn do_redis_code(url: &str, ca_file: &str, cert_file: &str, key_file: &str) -> redis::RedisResult<()> {
+///     let cert_file = File::open(cert_file).expect("cannot open private cert file");
+///     let mut reader_cert = BufReader::new(cert_file);
+///
+///     let key_file = File::open(key_file).expect("Cannot open private key file");
+///     let mut reader_key = BufReader::new(key_file);
+///
+///     let ca_file = File::open(ca_file).expect("Cannot open CA cert file");
+///     let mut reader_ca = BufReader::new(ca_file);
+///
+///     let client = build_with_tls(
+///         url,
+///         &mut reader_cert,
+///         &mut reader_key,
+///         &mut reader_ca
+///     )
+///     .expect("Unable to build client");
+///
+///     let connection_info = client.get_connection_info();
+///
+///     println!(">>> connection info: {connection_info:?}");
+///
+///     let mut con = client.get_async_connection().await?;
+///
+///     con.set("key1", b"foo").await?;
+///
+///     redis::cmd("SET")
+///         .arg(&["key2", "bar"])
+///         .query_async(&mut con)
+///         .await?;
+///
+///     let result = redis::cmd("MGET")
+///         .arg(&["key1", "key2"])
+///         .query_async(&mut con)
+///         .await;
+///     assert_eq!(result, Ok(("foo".to_string(), b"bar".to_vec())));
+///     println!("Result from MGET: {result:?}");
+///
+///     Ok(())
+/// }
+/// ```
+pub fn build_with_tls<C: IntoConnectionInfo>(
```

Comment:
I think this might make more sense as a method of `Client`, since it's basically a constructor, right?

Reply:
it's a constructor aye, but kept it here to avoid setting lot of attributes to condition the compilation to only when `rustls` is enabled. 
I could use an internal build method in `tls.rs` and call from a constructor wrapper in the client that would only be defined if the feature is available. Wdyt?

Reply:
yeah, the idea is add a constructor, but use int method to avoid lots of directives to dependencies conditioned to `rustls`: `retrieve_tls_certificates`, `ConnectionAddr`, `RedisError` and `ErrorKind`...

Reply:
done